src/main/java/cubicchunks/world/column/CubeMap.java

```diff
@@ -23,77 +23,135 @@
+	private final List<Cube> cubes = new ArrayList<>();
 
-	Cube get(int cubeY) {
-		return map.get(cubeY);
-	}
+	private ExtendedBlockStorage[] toBlockTick = new ExtendedBlockStorage[0];
 
 	Cube remove(int cubeY) {
-		Cube cube = this.map.remove(cubeY);
-		if(cube != null) {
-			set.remove(cube.getStorage());
-		}
-		return cubeMap.remove(cubeY);
+		int index = binarySearch(cubeY);
+		return index < cubes.size() && cubes.get(index).getY() == cubeY ? cubes.remove(index) : null;
 	}
 
-	boolean put(int cubeY, Cube cube) {
+	void put(Cube cube) {
 		if (cube == null) {
 			throw new NullPointerException();
 		}
-		if (this.contains(cubeY)) {
-			throw new IllegalArgumentException("Cube at " + cubeY + " already exists!");
+		if (this.contains(cube.getY())) {
+			throw new IllegalArgumentException("Cube at " + cube.getY() + " already exists!");
 		}
-		this.cubeMap.put(cubeY, cube);
-		map.put(cubeY, cube);
-		set.add(cube.getStorage()); //TODO: Don't force Cube's to have an ExtendedBlockStorage (empty Cube's eat memory)
-		return true;
+		cubes.add(cube);
+		cubes.sort(ORDER); // kind of expensive... but puts don't happen much (wish there was a SortedArrayList)
 	}
 
-	//TODO: optimize this... iterating over a map is kinda slow
 	Iterable<Cube> cubes(int startY, int endY) {
-		if(startY > endY) {
-			return this.cubeMap.subMap(endY, true, startY, true).descendingMap().values();
+		boolean reverse = false;
+		if(startY > endY){
+			int i = startY;
+			startY = endY;
+			endY = i;
+			reverse = true;
+		}
+
+		int bottom = binarySearch(startY);
+		int top = binarySearch(endY + 1); // subList()'s second arg is exclusive so we need to add 1
+
+		if(bottom < cubes.size() && top <= cubes.size()){
+			return reverse ? Lists.reverse(cubes.subList(bottom, top)) : cubes.subList(bottom, top);
+		}else{
+			return Collections.emptyList();
 		}
-		return this.cubeMap.subMap(startY, true, endY, true).values();
 	}
 
 	private boolean contains(int cubeY) {
-		return this.map.containsKey(cubeY);
+		int index = binarySearch(cubeY);
+		return index < cubes.size() && cubes.get(index).getY() == cubeY;
 	}
 
-	//TODO: optimize this... iterating over a map is so slow! (and this is used for ticking chunks :/)
 	@Override public Iterator<Cube> iterator() {
-		return Iterators.unmodifiableIterator(this.map.values().iterator());
+		return cubes.iterator();
 	}
 
 	public Collection<Cube> all() {
-		return Collections.unmodifiableCollection(this.map.values());
+		return Collections.unmodifiableCollection(cubes);
 	}
 
 	public boolean isEmpty() {
-		return this.cubeMap.isEmpty();
+		return cubes.isEmpty();
 	}
 
-	ExtendedBlockStorage[] getStorageArrays() {
-		//TODO: automatically update set if there is a Cube with a new ExtendedBlockStorage
-		return set.getArray(); 
+	/**
+	 * @return An array of EBS's form Cubes that need ticking... kind of a hack but vanilla needs it
+	 */
+	ExtendedBlockStorage[] getStoragesToTick() {
+		if(!isToTickValid()) {
```

Comment:
getStoragesToTick() basicly gets a list of Cubes that need ticking (That is what vanilla uses it for)...
Should I make it so that the array is updated when a cube gets a ticket that requests ticking?
(right now it does it dynamicly)